File: api/reviews.py
import json, os, base64
from http.server import BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs
from urllib.request import urlopen, Request
from urllib.error import HTTPError
from datetime import datetime

GITHUB_TOKEN = os.environ.get('GITHUB_TOKEN', '')
REPO = 'almirasultabova/centr-vremya-posts'
BRANCH = 'main'

def gh_get(path):
    url = f'https://api.github.com/repos/{REPO}/contents/{path}?ref={BRANCH}'
    req = Request(url, headers={'Authorization': f'token {GITHUB_TOKEN}', 'Accept': 'application/vnd.github.v3+json'})
    try:
        with urlopen(req) as r:
            return json.loads(r.read())
    except HTTPError:
        return None

def gh_put(path, content_str, sha=None):
    url = f'https://api.github.com/repos/{REPO}/contents/{path}'
    body = {
        'message': f'Update {path}',
        'content': base64.b64encode(content_str.encode()).decode(),
        'branch': BRANCH
    }
    if sha:
        body['sha'] = sha
    req = Request(url, data=json.dumps(body).encode(), headers={
        'Authorization': f'token {GITHUB_TOKEN}',
        'Content-Type': 'application/json',
        'Accept': 'application/vnd.github.v3+json'
    }, method='PUT')
    with urlopen(req) as r:
        return json.loads(r.read())

def load_reviews(month):
    data = gh_get(f'reviews/{month}.json')
    if not data:
        return {}, None
    content = base64.b64decode(data['content']).decode()
    return json.loads(content), data['sha']

class handler(BaseHTTPRequestHandler):
    def log_message(self, format, *args):
        pass

    def send_json(self, code, obj):
        body = json.dumps(obj, ensure_ascii=False).encode()
        self.send_response(code)
        self.send_header('Content-Type', 'application/json; charset=utf-8')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Content-Length', len(body))
        self.end_headers()
        self.wfile.write(body)

    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

    def do_GET(self):
        parsed = urlparse(self.path)
        params = parse_qs(parsed.query)
        month = params.get('month', ['june'])[0]
        reviews, _ = load_reviews(month)
        self.send_json(200, reviews)

    def do_POST(self):
        length = int(self.headers.get('Content-Length', 0))
        body = json.loads(self.rfile.read(length))
        month = body.get('month', 'june')
        idx = str(body.get('idx'))

        reviews, sha = load_reviews(month)
        if idx not in reviews:
            reviews[idx] = {}
        if 'status' in body:
            reviews[idx]['status'] = body['status']
        if 'comment' in body:
            reviews[idx]['comment'] = body['comment']
        reviews[idx]['updated'] = datetime.now().strftime('%d.%m %H:%M')

        gh_put(f'reviews/{month}.json', json.dumps(reviews, ensure_ascii=False, indent=2), sha)
        self.send_json(200, {'ok': True})

File: api/reviews.js
const https = require('https');

const GITHUB_TOKEN = process.env.GITHUB_TOKEN;
const REPO = 'almirasultabova/centr-vremya-posts';
const BRANCH = 'main';

function ghGet(path) {
  return new Promise((resolve, reject) => {
    const req = https.request({
      hostname: 'api.github.com',
      path: `/repos/${REPO}/contents/${path}?ref=${BRANCH}&t=${Date.now()}`,
      headers: { 'Authorization': `token ${GITHUB_TOKEN}`, 'Accept': 'application/vnd.github.v3+json', 'User-Agent': 'centr-vremya-posts' }
    }, res => {
      let d = ''; res.on('data', c => d += c); res.on('end', () => resolve({ status: res.statusCode, body: d }));
    });
    req.on('error', reject); req.end();
  });
}

function ghPut(path, body) {
  const data = JSON.stringify(body);
  return new Promise((resolve, reject) => {
    const req = https.request({
      hostname: 'api.github.com',
      path: `/repos/${REPO}/contents/${path}`,
      method: 'PUT',
      headers: {
        'Authorization': `token ${GITHUB_TOKEN}`,
        'Accept': 'application/vnd.github.v3+json',
        'User-Agent': 'centr-vremya-posts',
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(data)
      }
    }, res => {
      let d = ''; res.on('data', c => d += c); res.on('end', () => resolve({ status: res.statusCode, body: d }));
    });
    req.on('error', reject); req.write(data); req.end();
  });
}

function parseBody(req) {
  return new Promise((resolve, reject) => {
    let data = '';
    req.on('data', chunk => data += chunk);
    req.on('end', () => {
      try { resolve(JSON.parse(data)); } catch(e) { resolve({}); }
    });
    req.on('error', reject);
  });
}

async function loadReviews(month) {
  const res = await ghGet(`reviews/${month}.json`);
  if (res.status !== 200) return { data: {}, sha: null };
  const json = JSON.parse(res.body);
  const content = Buffer.from(json.content.replace(/\n/g, ''), 'base64').toString('utf8');
  return { data: JSON.parse(content), sha: json.sha };
}

module.exports = async (req, res) => {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  if (req.method === 'OPTIONS') { res.status(200).end(); return; }

  if (req.method === 'GET') {
    const month = (req.query && req.query.month) || 'june';
    try {
      const { data } = await loadReviews(month);
      res.status(200).json(data);
    } catch(e) {
      res.status(500).json({ error: e.message });
    }
    return;
  }

  if (req.method === 'POST') {
    try {
      const body = await parseBody(req);
      const month = body.month || 'june';
      const incomingReviews = body.reviews || {};

      const { sha } = await loadReviews(month);
      const content = Buffer.from(JSON.stringify(incomingReviews, null, 2)).toString('base64');
      const payload = { message: `update reviews/${month}.json`, content, branch: BRANCH };
      if (sha) payload.sha = sha;

      const result = await ghPut(`reviews/${month}.json`, payload);
      if (result.status >= 400) {
        res.status(result.status).json({ error: 'GitHub write failed', detail: result.body });
      } else {
        res.status(200).json({ ok: true });
      }
    } catch(e) {
      res.status(500).json({ error: e.message });
    }
    return;
  }

  res.status(405).json({ error: 'Method not allowed' });
};
